open_lm/file_utils.py

```diff
@@ -194,28 +269,84 @@ def get_string_for_epoch(num_samples, starting_chunk, paths, weights, min_shards
+                "reuse of samples."
             )
 
     for i, source_path in enumerate(paths):
         shard_list_source = shard_list_per_source[i]
         num_samples_source = num_samples_per_source[i]
         shard_root_source = "/".join(source_path.split("/")[:-1]) + "/"
-        shard_string_source = shard_root_source + "{" + ",".join(shard_list_source) + "}.tar"
+        if len(shard_list_source) == 1:
+            shard_string_source = shard_root_source + shard_list_source[0] + ".tar"
+        else:
+            shard_string_source = shard_root_source + "{" + ",".join(shard_list_source) + "}.tar"
         if source_path.startswith("s3"):
             shard_string_source = f"pipe:aws s3 cp {shard_string_source} -"
         shard_strings_per_source.append(shard_string_source)
 
     return shard_strings_per_source, num_samples_per_source, next_chunk
 
 
-if __name__ == "__main__":
-    print(sys.argv)
-    # path = 's3://s-laion/open_lm_shuffle/rpj_shuffled_mosaic_upsampled_tiktoken_100k_shards_try2/shard_metadata.jsonl'
-    # out = get_string_for_epoch(10000000000 // 2048, 10, path)
+def _single_epoch_string(num_samples, starting_shard_per_source, paths, weights, num_workers, world_size):
+    if len(paths) > 1:
+        logging.warning(
+            "Multiple sources are not supported fully as of now. Best effort will be done to mix the sources with "
+            "the desired ratio."
+        )
+
+    if weights is None:
+        weights = [1.0 / len(paths) for _ in range(len(paths))]
+    needed_samples_per_source = [int(np.ceil(weights[i] * num_samples / sum(weights))) for i in range(len(weights))]
+
+    manifests = [get_metadata_file(path) for path in paths]
+    shard_strings_per_source = []
+    next_shard_per_source = copy.deepcopy(starting_shard_per_source)
+    shard_list_per_source = [[] for _ in range(len(paths))]
+    num_samples_per_source = [0 for _ in range(len(paths))]
 
+    total_num_workers = num_workers * world_size
+    while not enough_shards(shard_list_per_source, total_num_workers) or not enough_samples(
+        num_samples_per_source, needed_samples_per_source
+    ):
+        try:
+            for i in range(len(paths)):
+                # Add shards incrementally
+                shard_list_per_source[i].append(manifests[i][next_shard_per_source[i]]["shard"])
+                num_samples_per_source[i] += manifests[i][next_shard_per_source[i]]["num_chunks"]
+                next_shard_per_source[i] += 1
+        except IndexError as e:
+            logging.error(
+                "Number of shards requested for a single epoch is more than the number of shards available. "
+                "Consider lowering the number of workers and / or the number of GPUs."
+            )
+            raise e
+
+    for i in range(len(paths)):
+        # Have the number of shards be a multiple of the total workers.
+        num_multiples = len(shard_list_per_source[i]) // total_num_workers
```

Comment:
One follow up - I think I don't understand why we need the number of shards to to be a multiple of total workers. Isn't it okay if some workers have more shards than others, since the number of samples per shard may vary anyway?

Reply:
This is mostly to avoid throwing away shards - the way webdataset works, each worker goes through a predetermined list of shards (so it cannot say get samples from shards assigned to a different worker or a different node.

We need at least one shard per worker, and it wouldn't cause trouble if we assigned more - but I think this is more convenient (since the leftover shards can be used in the next checkpoint).

Reply:
But with webdataset defaults, some workers can have more shards than others, right? If so, and the number of samples per shard varies, it's not clear to me that ensuring the number of shards is a multiple of number of workers is the right strategy. We could have some imbalanced shard assignments that end up working out in our favor when we adjust_samples, for example [*]. I might be missing something though, please let me know if so!

What we might want instead is to remove shards if one of the workers is going to ignore it completely - but we won't know this until after we call adjust_samples. 

[*] Suppose we have 5 shards with the following number of samples each: [1, 2, 1, 3, 3], and 2 workers. In this case, I believe WDS would assign shards 0, 2, and 4 to worker 1, and shards 1 and 3 to worker 2. If so, ensuring num_shards is a multiple of num_workers would actually lead to suboptimal results: we would save the last shard for the next epoch, and worker 2 would have to drop samples from shard 3 in adjust_samples to match the number of samples in worker 1.

Reply:
Yes, webdataset won't complain if you have an imbalanced shard assignment between workers, and I agree that in the above example rounding shards down is not the best thing to do. However, I think the way tokenization works currently most shards have the same size - so this scenario isn't as likely.

I agree that the ideal solution is to verify which shards will be used and which ones won't be. However, it might be cleaner if we just have this requirement (even if suboptimal in some cases).

Reply:
Okay, makes sense. Could you leave a short comment mentioning this motivation? Something like this, or even more concise if you can make it?

```
# Ensure the number of shards is a multiple of number of workers, so each worker has the same
# number of shards.
#
# This is a heuristic to minimize how much data we discard when trying to ensure each worker has
# the same number of samples. Shards tend to have similar number of samples, so an extra shard
# in a worker will likely get discarded.
```

Reply:
Sounds good - will do that!

Reply:
On a related note, after our discussion offline it seems it's best to not use `adjust_samples` - setting it to the maximum number of samples is pointless. It's best to just let the workers run out of data.

Reply:
Could we through a warning if the manifest has highly imbalanced shards? The way the tokenize+shuffle code currently works the shard sizes should be approximately equal.

Reply:
I included this check as well. We will throw a warning if a source contains small shards with respect to itself + if sources are imbalanced.